File: mobile/test/modules/merchant_claim/merchant_claim_evidence_policy_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:tum2/modules/merchant_claim/models/merchant_claim_evidence_policy.dart';
import 'package:tum2/modules/merchant_claim/models/merchant_claim_models.dart';

void main() {
  test('policy farmacia exige evidencia regulatoria', () {
    final policy = resolveMerchantClaimEvidencePolicy('farmacia');
    expect(policy.categoryId, 'farmacia');
    expect(
      policy.requiredAdditionalKinds,
      contains(MerchantClaimEvidenceKind.regulatoryDocument),
    );
  });

  test('policy comida_al_paso acepta visual y vínculo contextual', () {
    final policy = resolveMerchantClaimEvidencePolicy('comida_al_paso');
    final evidence = [
      const MerchantClaimEvidenceFile(
        id: 'e1',
        kind: MerchantClaimEvidenceKind.operationalPointPhoto,
        storagePath: 'merchant-claims/u1/c1/operational_point_photo/a.jpg',
        contentType: 'image/jpeg',
        sizeBytes: 900,
      ),
      const MerchantClaimEvidenceFile(
        id: 'e2',
        kind: MerchantClaimEvidenceKind.alternativeRelationshipEvidence,
        storagePath:
            'merchant-claims/u1/c1/alternative_relationship_evidence/b.jpg',
        contentType: 'image/jpeg',
        sizeBytes: 1100,
      ),
    ];
    expect(policy.isSatisfied(evidence), isTrue);
  });

  test('fallback fuerza revisión manual por categoría desconocida', () {
    final policy = resolveMerchantClaimEvidencePolicy('legacy_unknown');
    expect(policy.categoryId, 'fallback');
    expect(policy.manualReviewTriggers,
        contains('fallback_category_policy_applied'));
  });

  test('categorías canónicas resuelven a policy canónica', () {
    expect(
        resolveMerchantClaimEvidencePolicy('farmacia').categoryId, 'farmacia');
    expect(
      resolveMerchantClaimEvidencePolicy('veterinaria').categoryId,
      'veterinaria',
    );
    expect(
      resolveMerchantClaimEvidencePolicy('comida_al_paso').categoryId,
      'comida_al_paso',
    );
  });

  test('allowlist de claims bloquea categorías no MVP', () {
    expect(isAllowedMerchantClaimCategoryId('panaderia'), isTrue);
    expect(isAllowedMerchantClaimCategoryId('bakery'), isFalse);
    expect(isAllowedMerchantClaimCategoryId('confiteria'), isTrue);
    expect(isAllowedMerchantClaimCategoryId('otro'), isFalse);
    expect(isAllowedMerchantClaimCategoryId('farmacia'), isTrue);
    expect(isAllowedMerchantClaimCategoryId('kiosco'), isTrue);
  });
}

File: functions/src/lib/__tests__/merchantClaimEvidencePolicy.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import {
  CLAIM_EVIDENCE_POLICY_VERSION,
  evaluateEvidenceAgainstPolicy,
  isAllowedClaimCategoryId,
  normalizeClaimCategoryId,
  resolveClaimEvidencePolicy,
} from "../merchantClaimEvidencePolicy";

test("policy resolver devuelve strict para farmacia", () => {
  const policy = resolveClaimEvidencePolicy("farmacia");
  assert.equal(policy.policyVersion, CLAIM_EVIDENCE_POLICY_VERSION);
  assert.equal(policy.strictnessLevel, "regulated_strict");
  assert.deepEqual(policy.requiredAdditionalEvidence, ["regulatory_document"]);
});

test("categorías canónicas resuelven policy canónica", () => {
  const farmacia = resolveClaimEvidencePolicy("farmacia");
  const veterinaria = resolveClaimEvidencePolicy("veterinaria");
  const comidaAlPaso = resolveClaimEvidencePolicy("comida_al_paso");
  assert.equal(farmacia.categoryId, "farmacia");
  assert.equal(veterinaria.categoryId, "veterinaria");
  assert.equal(comidaAlPaso.categoryId, "comida_al_paso");
});

test("categorías no MVP quedan fuera de allowlist de claims", () => {
  assert.equal(normalizeClaimCategoryId("panaderia"), "panaderia");
  assert.equal(normalizeClaimCategoryId("bakery"), "bakery");
  assert.equal(normalizeClaimCategoryId("confiteria"), "confiteria");
  assert.equal(normalizeClaimCategoryId("cafeteria"), "cafeteria");
  assert.equal(isAllowedClaimCategoryId("bakery"), false);
  assert.equal(isAllowedClaimCategoryId("cafeteria"), false);
  assert.equal(isAllowedClaimCategoryId("panaderia"), true);
  assert.equal(isAllowedClaimCategoryId("confiteria"), true);
  assert.equal(isAllowedClaimCategoryId("farmacia"), true);
});

test("fallback policy evita auto-aprobación implícita para categoría desconocida", () => {
  const evaluation = evaluateEvidenceAgainstPolicy({
    categoryId: "legacy_unknown_category",
    evidenceKinds: new Set(["storefront_photo", "ownership_document"]),
  });
  assert.equal(evaluation.requiredEvidenceSatisfied, true);
  assert.equal(evaluation.sufficiencyLevel, "sufficient_manual_review");
  assert.ok(evaluation.manualReviewReasons.includes("fallback_category_policy_applied"));
});

test("comida_al_paso acepta combinación flexible operacional + alternativa", () => {
  const evaluation = evaluateEvidenceAgainstPolicy({
    categoryId: "comida_al_paso",
    evidenceKinds: new Set([
      "operational_point_photo",
      "alternative_relationship_evidence",
    ]),
  });
  assert.equal(evaluation.requiredEvidenceSatisfied, true);
  assert.equal(evaluation.missingEvidenceTypes.length, 0);
});

test("veterinaria exige evidencia reforzada específica", () => {
  const evaluation = evaluateEvidenceAgainstPolicy({
    categoryId: "veterinaria",
    evidenceKinds: new Set(["storefront_photo", "ownership_document"]),
  });
  assert.equal(evaluation.requiredEvidenceSatisfied, false);
  assert.ok(evaluation.missingEvidenceTypes.includes("reinforced_relationship_evidence"));
});
